### PR TITLE
Shutdown Python REPL correctly

### LINKED ISSUE #18409: Abort on REPL shutdown due to live objects
#### Problem
When shutting down the Matter Controller I get an abort with the following stack trace:

```cpp
#0  0x00007ffff7a7620c in __pthread_kill_implementation () at /usr/lib/libc.so.6
#1  0x00007ffff7a26838 in raise () at /usr/lib/libc.so.6
#2  0x00007ffff7a10535 in abort () at /usr/lib/libc.so.6
#3  0x00007fffef0e8d07 in chipAbort() () at ../../src/lib/support/CodeUtils.h:489
#4  0x00007fffef242bc1 in chip::HeapObjectPool<chip::Inet::UDPEndPointImplSockets>::~HeapObjectPool() (this=0x7fffef4f9f18 <chip::DeviceLayer::UDPEndPointManager()::gUDPEndPointManager+24>, __in_chrg=<optimized out>) at ../../src/lib/support/Pool.h:317
#5  0x00007fffef242636 in chip::ObjectPool<chip::Inet::UDPEndPointImplSockets, 32ul, (chip::ObjectPoolMem)1>::~ObjectPool() (this=0x7fffef4f9f18 <chip::DeviceLayer::UDPEndPointManager()::gUDPEndPointManager+24>, __in_chrg=<optimized out>) at ../../src/lib/support/Pool.h:451
#6  0x00007fffef242d98 in chip::Inet::EndPointManagerImplPool<chip::Inet::UDPEndPointImplSockets>::~EndPointManagerImplPool() (this=0x7fffef4f9f00 <chip::DeviceLayer::UDPEndPointManager()::gUDPEndPointManager>, __in_chrg=<optimized out>) at ../../src/inet/InetLayer.h:121
#7  0x00007ffff7a28ef5 in __run_exit_handlers () at /usr/lib/libc.so.6
#8  0x00007ffff7a29070 in on_exit () at /usr/lib/libc.so.6
#9  0x00007ffff7a11297 in __libc_start_call_main () at /usr/lib/libc.so.6
#10 0x00007ffff7a1134a in __libc_start_main_impl () at /usr/lib/libc.so.6
#11 0x0000555555555055 in _start ()
```

From what I understand it's just an assertion which checks if every instance Object in the Pool has been released. So it seems that not all EndPoints allocated through the `UDPEndPointManager` have been released?

```cpp
#if __SANITIZE_ADDRESS__
        // Free all remaining objects so that ASAN can catch specific use-after-free cases.
        ReleaseAll();
#else  // __SANITIZE_ADDRESS__
       // Verify that no live objects remain, to prevent potential use-after-free.
        VerifyOrDie(Allocated() == 0);
#endif // __SANITIZE_ADDRESS__
```

Also, it does seem that when Clang's address sanitizer is in use the code de-allocate all objects using `ReleaseAll()`.

This has been observed when using the Python Controller and minimal mDNS implementation.

Anyone else get this or is this specific to my setup?  Is this something worth tracking down?

 #### Proposed Solution
Not sure.

### PR DESCRIPTION
**Pull request by agners**
#### Problem
* Fix crash when REPL is being shutdown
* Fixes #18409

#### Change overview
Uses `atexit` module to Shutdown the CHIP stack when they Python REPL quits. Also addresses an issue where the Storage manager tries to access the chipStack after Shutdown.

#### Testing
Using `chip-repl` in dev environment, quit while debugger is attached. No more segfaults.

### CODE REVIEW

File: src/controller/python/chip/ChipStack.py [hunk 1]
```diff
@@ -318,6 +318,7 @@ def setLogFunct(self, logFunct):
             self._ChipStackLib.pychip_Stack_SetLogFunct(logFunct)
 
     def Shutdown(self):
+        self._persistentStorage = None
```
Comment (mrjerryjohns) [inline src/controller/python/chip/ChipStack.py:321]: So, what exactly happens if this line is not added?

**Review by mrjerryjohns — APPROVED** [review]

File: src/controller/python/chip/ChipStack.py (hunk 1 above)
Comment (agners) [reply src/controller/python/chip/ChipStack.py:321]: We later fail to access the builtins.chipStack:
AttributeError: module 'builtins' has no attribute 'chipStack'

Comment (agners) [reply src/controller/python/chip/ChipStack.py:321]: From what I understand, `PersistentStorage.__del__` is called much later, presumably after `builtins.chipStack` has been cleaned up?

```
2022-05-13 18:48:30 allenwind chip.DL[974255] ERROR System Layer shutdown                                                                                                                                                                                                                                                     
[Thread 0x7ffff57d0640 (LWP 974259) exited]                                                                                                                                                                                                                                                                                   
Exception ignored in: <function PersistentStorage.__del__ at 0x7ffff4b8fca0>                                                                                                                                                                                                                                                  
Traceback (most recent call last):                                                                                                                                                                                                                                                                                            
  File "/home/sag/projects/project-chip/connectedhomeip/out/python_env/lib/python3.9/site-packages/chip/storage/__init__.py", line 157, in __del__
AttributeError: module 'builtins' has no attribute 'chipStack'
```

Comment (mrjerryjohns) [reply src/controller/python/chip/ChipStack.py:321]: By setting it to None, is this then re-ordering the destruction order such that PersistentStorage is getting deleted first before ChipStack?

Comment (agners) [reply src/controller/python/chip/ChipStack.py:321]: > By setting it to None, is this then re-ordering the destruction order such that PersistentStorage is getting deleted first before ChipStack?

Yes, that is exactly what is happening, before this change:

```
In [1]: quit
[Thread 0x7fffdf7fe640 (LWP 1452570) exited]
2022-05-18 15:51:03 allenwind chip.BLE[1452557] ERROR BleConnectionDelegate::CancelConnection is not implemented.
2022-05-18 15:51:03 allenwind chip.DL[1452557] ERROR renamed tmp file to file (/tmp/chip_counters.ini)
2022-05-18 15:51:03 allenwind chip.DL[1452557] ERROR Inet Layer shutdown
2022-05-18 15:51:03 allenwind chip.DL[1452557] ERROR BLE shutdown
2022-05-18 15:51:03 allenwind chip.DL[1452557] ERROR System Layer shutdown
[Thread 0x7ffff57cf640 (LWP 1452562) exited]
Destroying storage
Exception ignored in: <function PersistentStorage.__del__ at 0x7ffff4aefe50>
Traceback (most recent call last):
  File "/home/sag/projects/project-chip/connectedhomeip/out/python_env/lib/python3.9/site-packages/chip/storage/__init__.py", line 159, in __del__
AttributeError: module 'builtins' has no attribute 'chipStack'
```

After this change:

```
In [1]: quit
[Thread 0x7fffdf7fe640 (LWP 1453211) exited]
Destroying storage
2022-05-18 15:52:31 allenwind chip.BLE[1453201] ERROR BleConnectionDelegate::CancelConnection is not implemented.
2022-05-18 15:52:31 allenwind chip.DL[1453201] ERROR renamed tmp file to file (/tmp/chip_counters.ini)
2022-05-18 15:52:31 allenwind chip.DL[1453201] ERROR Inet Layer shutdown
2022-05-18 15:52:31 allenwind chip.DL[1453201] ERROR BLE shutdown
2022-05-18 15:52:31 allenwind chip.DL[1453201] ERROR System Layer shutdown
[Thread 0x7ffff57cf640 (LWP 1453204) exited]
[Thread 0x7fffdffff640 (LWP 1453210) exited]
[Thread 0x7fffecb3a640 (LWP 1453209) exited]
[Thread 0x7fffed33b640 (LWP 1453208) exited]
[Thread 0x7fffedb3c640 (LWP 1453207) exited]
[Thread 0x7fffee33d640 (LWP 1453206) exited]
[Inferior 1 (process 1453201) exited normally]
```


I'll add a comment.

**Pull request merged**